### PR TITLE
[EWT-192] Add support for a limited set of custom HTTP headers

### PR DESCRIPTION
**Pull request by dili91**
# Description

We're proposing a new feature that will give customers the ability to _optionally_ set the following HTTP headers: 
- `X-Forwarded-For`
- `Tl-Signature`
- `Idempotency-Key`

In the new version of the library (`8.0.0`) for each dispositive method we'll add an overload that will include an `HttpHeaders` parameter. This means that the following 2 methods will be available for creating a payment, as an example: 

```java
    CompletableFuture<ApiResponse<CreatePaymentResponse>> createPayment(CreatePaymentRequest request);

    CompletableFuture<ApiResponse<CreatePaymentResponse>> createPayment(Headers headers, CreatePaymentRequest request);
```

The new header object will be built with the help of a builder: 
```java
var headers = Headers.builder()
        .idempotencyKey("a-custom-key")
        .signature("a-custom-signature")
        .xForwardedFor("1.2.3.4")
        .build();
```
## Breaking change notice

This is going to be a breaking change for 2 reasons: 
- we won't expose APIs interfaces anymore, but _handler_ types only
- the old mechanism to set the `X-Forwarded-For` header on the `startAuthorizationFlow` invocations is superseded by the new API. 
- The (empty) SubmitConsentRequest object has been removed in favour of a more generic empty request body type.

## Type of change

Please select multiple options if required.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code where necessary
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the relevant documentation

### CODE REVIEW

File: src/main/java/com/truelayer/java/payments/entities/SubmitConsentRequest.java
Comment (dili91) [inline src/main/java/com/truelayer/java/payments/entities/SubmitConsentRequest.java:1]: removed in favour of a generic Empty request body type

File: src/main/java/com/truelayer/java/http/mappers/HeadersMapper.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import com.truelayer.java.Constants;
+import com.truelayer.java.http.entities.Headers;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Utility class for Headers to Map conversions.
+ */
+public class HeadersMapper {
```
Comment (dili91) [inline src/main/java/com/truelayer/java/http/mappers/HeadersMapper.java:15]: I would have loved to use the default visibility here but I would need to restructure a lot.

File: src/main/java/com/truelayer/java/payments/IPaymentsApi.java [hunk 2]
```diff
@@ -15,15 +15,16 @@
  * @see <a href="https://docs.truelayer.com/reference/create-payment"><i>Payments</i> API reference</a>
  */
 public interface IPaymentsApi {
```
Comment (dili91) [inline src/main/java/com/truelayer/java/payments/IPaymentsApi.java:17]: I would have loved to use the `default` visibility here but I would need to restructure a lot.

File: src/main/java/com/truelayer/java/TrueLayerClientBuilder.java [hunk 3]
```diff
@@ -246,7 +255,7 @@ public TrueLayerClient build() {
                 merchantAccountsHandler,
                 mandatesHandler,
                 payoutsHandler,
-                hppLinkBuilder,
-                commonApiHandler);
+                commonHandler,
+                hppLinkBuilder);
```
[the remark this answers is not in the document]
Comment (dili91) [reply src/main/java/com/truelayer/java/TrueLayerClientBuilder.java:259]: I've also changed the order [here](https://github.com/TrueLayer/truelayer-java/pull/204/files#diff-00f7625b18f45545b8b3ad623091cd90fc84f442bfdea8120d92d57d0ff156fcL32) :) 

Wanted to keep the link builder as last as its conceptually different

**Review by tl-luca-baggi — APPROVED** [review]

File: src/main/java/com/truelayer/java/TrueLayerClient.java [hunk 4]
```diff
@@ -23,22 +23,24 @@
...
+    private ICommonHandler commonHandler;
+
     private IHostedPaymentPageLinkBuilder hostedPaymentPageLinkBuilder;
-    private ICommonApi commonApi;
 
     public TrueLayerClient(
             IAuthenticationHandler authenticationHandler,
             IHostedPaymentPageLinkBuilder hostedPaymentPageLinkBuilder,
-            ICommonApi commonApi) {
+            ICommonHandler commonApi) {
```
Comment (lcheung-at-square) [inline src/main/java/com/truelayer/java/TrueLayerClient.java:40]: Nit: Looks like this parameter's name wasn't modified to match its new type.

**Review by lcheung-at-square — APPROVED** [review]

File: src/main/java/com/truelayer/java/TrueLayerClient.java (hunk 4 above)
Comment (dili91) [reply src/main/java/com/truelayer/java/TrueLayerClient.java:40]: Thank you !

**Pull request merged**